pkg/virt-handler/isolation/isolation.go

```diff
@@ -215,34 +216,63 @@ func getMemlockSize(vm *v1.VirtualMachineInstance) (int64, error) {
 	return bytes_, nil
 }
 
-func NewIsolationResult(pid int, slice string, controller []string) *IsolationResult {
-	return &IsolationResult{pid: pid, slice: slice, controller: controller}
+func NewIsolationResult(pid int, slice string, controller []string) IsolationResult {
+	return &realIsolationResult{pid: pid, slice: slice, controller: controller}
 }
 
-type IsolationResult struct {
+type IsolationResult interface {
+	// cgroup slice
+	Slice() string
+	// process ID
+	Pid() int
+	// full path to the process namespace
+	PIDNamespace() string
+	// full path to the process root mount
+	MountRoot() string
+	// retrieve additional information about the process root mount
+	MountInfoRoot() (*MountInfo, error)
+	// full path to the mount namespace
+	MountNamespace() string
+	// full path to the network namespace
+	NetNamespace() string
+	// execute a function in the process network namespace
+	DoNetNS(func() error) error
+}
+
+type realIsolationResult struct {
 	pid        int
 	slice      string
 	controller []string
 }
 
-func (r *IsolationResult) Slice() string {
-	return r.slice
+func (r *realIsolationResult) DoNetNS(f func() error) error {
+	netns, err := ns.GetNS(r.NetNamespace())
+	if err != nil {
+		return fmt.Errorf("failed to get launcher pod network namespace: %v", err)
+	}
+	return netns.Do(func(_ ns.NetNS) error {
```

Comment:
Can you do `defer runtime.UnlockOSThread()` before this line and remove it from the callback in `vm.go`? Otherwise we would already unlock the thread before we swith back to the host namespace.

Reply:
The concern is legit, though I think I may need to do something slightly more advanced because Do() doesn't guarantee you the thread was ever locked, or the callback was ever called. But great catch, let me craft a clunky solution.

Reply:
ok that does not work. This netns implementation uses a go-routine internally which would make my suggestion useless.

Reply:
Since the namespace switch now only happens until the VM is running and since we have an increasing backoff in case of errors, It might be ok to just don't unlock right now and let the thread die. When @booxter and I were reconstructing  why the netns library is written that way it turns out that running the callback in an extra callback and not unlocking the thread is a very good and save decision to get a robust and correct result in all cases.

It is not that trivial to solve the thread unlock without modifying the vendored library. As discussed with @booxter we can then follow up with an improvement on the netns library to unlock the thread again in case of a success, or use a temp-file like @davidvossel  suggested.

@davidvossel what do you think?

Reply:
Pushed the branch without the original unlock patch as discussed. I will push another PR for `plugins` repo and see what their maintainers have to say to our suggestion.

Reply:
remove the unlock. lets merge the PR. then lets follow up with the tempfile option that marks locally when the network setup has completed.